### PR TITLE
feat(DS-635): connect GET endpoints 

### PR DESCRIPTION
**Pull request by bra-i-am**
## Description
This PR mainly applies the GET endpoints created in the new [Plugin CVW API](https://github.com/eduNEXT/platform-global-teacher-campus/pull/2) and involves the necessary changes for being able to use these endpoints where essential in the Frontend. Additionally, this PR involves the following: 
* feat: implement POST of validation processes
* feat: implement error validations with Yup for the form of submission
* refactor: allow selecting multiple categories
* refactor: correct some variables for proper and precise use

## How to test
1. Make sure you have installed the Plugin CVW and you're in the latest version of the ```api``` branch which makes reference to this [PR](https://github.com/eduNEXT/platform-global-teacher-campus/pull/2)
2. Go to the section [Platform_Global_Teacher_Campus](http://local.overhang.io:8000/admin/platform_global_teacher_campus/) in the admin panel and make sure to accomplish the following bullets (pre-requisites):
  - Go to [Course Categories](http://local.overhang.io:8000/admin/platform_global_teacher_campus/coursecategory/) and make sure you have at least one category, on the contrary, create one
  - Go to [Validation Bodies](http://local.overhang.io:8000/admin/platform_global_teacher_campus/validationbody/) and make sure you have at least one category, on the contrary, create one
  - You need to have a course available that does not have an opened validation_process, that is, the course is not in this [list](http://local.overhang.io:8000/admin/platform_global_teacher_campus/validationprocess/). If you have a course in this list, you can delete the validation process and keep going, or, on the other hand, you can go to the [STUDIO](http://studio.local.overhang.io:8001/home/) and create a new course 
 3. Go to the [Validation Panel](http://apps.local.overhang.io:2001/validation-panel) and access the modal for submitting a course for validation
 
image

TESTS:
1. You should see the **course** you have created in the selector 'Course Name'

Screenshot from 2023-09-18 16-46-17

2. You should see the **validation body** you have created in the selector 'Validation Body'
- [x] Connect GET /plugin-cvw/api/v1/validation-bodies

Screenshot from 2023-09-18 16-46-22

3. You should see the **categories** you have created in the selector 'Category'
- [x] Connect GET /plugin-cvw/api/v1/categories/

Screenshot from 2023-09-18 16-46-28

5. Once you select an option in every selector and write a comment, you'll be able to Submit the course and, after this, you are going to watch the table be updated with the validation process you just created
- [x] ( Additional ) Connect POST /plugin-cvw/api/v1/validation-processes/< COURSE_ID >/submit/ 
- [x] Connect GET /plugin-cvw/api/v1/validation-processes

image

6. And the last one, when you click the name of a course in the table, you are going to be able to see the information of the validation process with more details about the review and past processes
- [x] Connect GET /plugin-cvw/api/v1/validation-processes/< COURSE_ID >

image

### CODE REVIEW

File: src/data/slices/validationRecordSlice.js [hunk 1]
```diff
@@ -42,6 +44,29 @@ export const validationRecordSlice = createSlice({
...
+      state.availableValidationProcesses = {
+        ...state.availableValidationProcesses,
+        loadStatus: REQUEST_STATUS.LOADED,
+        error: {},
+      };
+    });
+    builder.addCase(postNewValidationProcess.rejected, (state, action) => {
+      state.availableValidationProcesses = {
+        ...state.availableValidationProcesses,
+        loadStatus: REQUEST_STATUS.LOADED,
```
Comment (dcoa) [inline src/data/slices/validationRecordSlice.js]:
```suggestion
        loadStatus: REQUEST_STATUS.FAILED,
```

Remember to check the latest changes in master branch.

File: src/data/slices/validationRecordSlice.js [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
-import { getValidationProcesses } from '../api';
+import { getValidationProcesses, postValidationProcess } from '../api';
 import { REQUEST_STATUS } from '../constants';
 
 export const getAvailableValidationProcesses = createAsyncThunk('organization/validationProcesses', async () => {
   const response = await getValidationProcesses();
   return response;
 });
 
+export const postNewValidationProcess = createAsyncThunk('organization/validationProcesses/Post', async (formData) => postValidationProcess(formData));
```
Comment (dcoa) [inline src/data/slices/validationRecordSlice.js]:
```suggestion
export const createValidationProcess = createAsyncThunk('organization/validationProcesses/post', async (formData) => postValidationProcess(formData));
```

Remember these are actions, so the names should be semantic with that.

File: src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx [hunk 3]
```diff
@@ -30,49 +37,66 @@ const CourseValidationRequestForm = ({ isOpen, close }) => {
...
-    handleChange, values, handleSubmit, setValues,
-  } = useFormik({
+  const FormSchema = Yup.object().shape({
+    courseId: Yup.string().required('Please select a course!'),
+    validationBodyId: Yup.number().required('Please select a validation body!'),
+    categoryIds: Yup.array().of(Yup.string()).min(1, 'Please select at least one category!'),
+    comment: Yup.string().required('Please insert at least a short description about your submission'),
+  });
+
+  const getAdaptedData = (formData) => ({
```
Comment (bra-i-am) [inline src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx]: Look for the best way of using this as a helper or a local state

File: src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx [hunk 4]
```diff
@@ -30,49 +37,66 @@ const CourseValidationRequestForm = ({ isOpen, close }) => {
...
+      availableCourseCategories.find(category => category.name === categoryName).id)),
+  });
+
+  const formik = useFormik({
     initialValues,
+    validationSchema: FormSchema,
     onSubmit: (formData) => {
-      console.log(formData);
+      dispatch(createValidationProcess(getAdaptedData(formData)));
+      dispatch(getAvailableValidationProcesses());
```
Comment (bra-i-am) [inline src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx]: Add handle close

File: src/data/constants.js [hunk 5]
```diff
@@ -1,11 +1,21 @@
 const VALIDATION_STATUS = {
```
Comment (bra-i-am) [inline src/data/constants.js:1]: Change this to an Object { key: VALIDATION_STATUS, label: VALIDATION_STATUS_LABEL }

File: src/components/Timeline/Timeline.jsx [hunk 6]
```diff
@@ -3,11 +3,11 @@ import RecordItem from './RecordItem';
 
 const Timeline = ({ pastProcessEvents, validationBody }) => (
```
Comment (bra-i-am) [inline src/components/Timeline/Timeline.jsx]: Send ```validationBody.name``` from the parent component

File: src/components/Timeline/Timeline.test.jsx [hunk 7]
```diff
@@ -39,8 +39,9 @@ describe('Timeline', () => {
     expect(container.querySelectorAll('li').length).toBe(2);
     expect(getByText(formatedMockDate, { exact: false })).toBeInTheDocument();
     expect(getByText(mockEvents[0].comment, { exact: false })).toBeInTheDocument();
+    screen.debug();
```
Comment (bra-i-am) [inline src/components/Timeline/Timeline.test.jsx]: Delete this

File: src/components/Timeline/Timeline.test.jsx [hunk 8]
```diff
@@ -13,15 +13,15 @@ const mockEvents = [
...
+    status: 'aprv',
     createdAt: '2023-08-20',
     reason: null,
     comment: 'This is the comment of Validator 2',
     user: 'Validator 2',
   },
 ];
 
-const mockData = { pastProcessEvents: mockEvents, validationBody: 'Validator Body #1' };
+const mockData = { pastProcessEvents: mockEvents, validationBody: { name: 'Validator Body #1' } };
```
Comment (bra-i-am) [inline src/components/Timeline/Timeline.test.jsx]: Change this to a string

File: src/components/ValidationProcess/FormLayout/FormLayout.jsx [hunk 9]
```diff
@@ -43,13 +43,17 @@ const FormLayout = ({
     .map(
       (field) => {
         const isDisabled = !isValidator || field.disabled;
+        const value = field.name.toLowerCase().includes('date')
```
Comment (bra-i-am) [inline src/components/ValidationProcess/FormLayout/FormLayout.jsx]: FInd a better way to render this

File: src/components/ValidationTable/ValidationTable.jsx [hunk 10]
```diff
@@ -142,7 +142,7 @@ const ValidationTable = ({ data, isLoading }) => {
       <ModalLayout
         isOpen={isOpen}
         onClose={close}
-        tabs={[
+        tabs={currentValidationRecord.loadStatus !== REQUEST_STATUS.LOADING && [
```
Comment (bra-i-am) [inline src/components/ValidationTable/ValidationTable.jsx]: Find the way of using a loader spinner

Comment (bra-i-am) [reply src/components/ValidationTable/ValidationTable.jsx]: isLoading in the ModalLayout.jsx

File: src/data/api.js [hunk 11]
```diff
@@ -5,11 +5,12 @@ import { getAuthenticatedHttpClient, getAuthenticatedUser } from '@edx/frontend-
 import { VALIDATION_API_PATH } from './constants';
 import mockedValidationProcesses from './mocked_data';
 
-ensureConfig(['LMS_BASE_URL'], 'Validation Panel API');
+ensureConfig(['LMS_BASE_URL', 'STUDIO_BASE_URL'], 'Validation Panel API');
```
Comment (bra-i-am) [inline src/data/api.js]: DELETE THIS

File: src/data/api.js [hunk 12]
```diff
@@ -44,62 +45,43 @@ export async function getValidationProcesses() {
...
 
 /**
- * Fetches all the categories associated with the current course.
- *  @param {string} courseId
+ * Fetches all available categories to associate with course validation processes
  * @returns {Promise<[{}]>}
  */
 
-export async function getCourseCategories(courseId) {
+export async function getAvailableCategories() {
```
Comment (bra-i-am) [inline src/data/api.js]: getCategories()

File: src/data/api.js [hunk 13]
```diff
@@ -113,7 +95,7 @@ export async function getCourseCategories(courseId) {
  *  */
```
Comment (bra-i-am) [inline src/data/api.js:89]: Update JSDoc

File: src/data/slices/categoriesSlice.js [hunk 14]
```diff
@@ -3,7 +3,7 @@ import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
 import { getAvailableCategories } from '../api';
 import { REQUEST_STATUS } from '../constants';
 
-export const getCourseCategoriesByCourseId = createAsyncThunk('courses/courseCategories', async () => {
+export const getAvailableCourseCategories = createAsyncThunk('courses/courseCategories', async () => {
```
Comment (bra-i-am) [inline src/data/slices/categoriesSlice.js]: getCourseCategories()

File: src/data/slices/currentValidationRecordSlice.js [hunk 15]
```diff
@@ -1,10 +1,19 @@
 /* eslint-disable no-param-reassign */
-import { createSlice } from '@reduxjs/toolkit';
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+import { getValidationProcess } from '../api';
+import { REQUEST_STATUS } from '../constants';
+
+export const getCurrentValidationProcess = createAsyncThunk('organization/validationProcesses/selected', async (courseId) => {
```
Comment (bra-i-am) [inline src/data/slices/currentValidationRecordSlice.js]: getCurrentValidationProcessByCourseId()

File: src/data/slices/currentValidationRecordSlice.js [hunk 16]
```diff
@@ -1,10 +1,19 @@
...
+  const response = await getValidationProcess(courseId);
+  return response;
+});
 
 const currentValidationRecordInitialState = {
-  loadStatus: null,
+  loadStatus: 'idle',
```
Comment (bra-i-am) [inline src/data/slices/currentValidationRecordSlice.js:12]: null

File: src/utils/helpers.js [hunk 17]
```diff
@@ -1,60 +1,68 @@
...
 
+  const [lastValidationProcessEvent] = getLastAndFirstValidationProcessEvents(course);
+
   return {
     isExempted,
     courseName: course.courseName,
     courseId: course.courseId,
-    organization: course.organization,
-    categories: course.categories,
+    reviewer: course.currentValidationUser || lastValidationProcessEvent?.user,
```
Comment (bra-i-am) [inline src/utils/helpers.js]: Left just the currentValidationUser

File: src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx (hunk 3 above)
Comment (bra-i-am) [reply src/components/CourseValidationRequestForm/CourseValidationRequestForm.jsx]: Implemented as a helper

**Comment by dcoa** [comment]
We have to be careful with the visitor experience I  just fill in the first field and display all the message errors for the subsequent inputs 

image

This validation should be displayed when the user presses the submit button not before. 

This could be resolved in a different PR understanding that this PR should be focused on GET endpoints

**Review by dcoa — APPROVED** [review]

**Pull request merged**